### PR TITLE
Postgres JSON nested field columns

### PR DESCRIPTION
**Pull request by howonlee**
Continue on long road to #708, aiming to unblock Kiselev. This actually links in all the syncing stuff we've been doing. JSON tables should now emit nested field column fields, which just act like normal fields but have `nfc_path` property that indicates they're nested field columns.

There needs to be a lot more manual and automated testing but I'm aiming at unblocking Kiselev so he can work on this one without the tests yet

### CODE REVIEW

**Comment by howonlee** [comment]
Steps to work on this on the frontend:
1. Use Postgres DB
2. Create DB, table with columns which are JSON type and, importantly, all the same JSON path structure
3. After a sync there should just be a bunch of new columns corresponding to the internal JSON bits

File: src/metabase/driver/postgres.clj [hunk 1]
```diff
@@ -32,6 +32,8 @@
 
 (driver/register! :postgres, :parent :sql-jdbc)
 
+(defmethod driver/supports? [:postgres :nested-field-columns] [_ _] true)
```
Comment (dpsutton) [inline src/metabase/driver/postgres.clj:35]: this is deprecated and we should prefer `database-supports?` going forward.

File: src/metabase/sync/sync_metadata/fields/fetch_metadata.clj [hunk 2]
```diff
@@ -94,4 +84,6 @@
   "Fetch metadata about Fields belonging to a given `table` directly from an external database by calling its driver's
   implementation of `describe-table`."
   [database :- i/DatabaseInstance table :- i/TableInstance]
-  (:fields (fetch-metadata/table-metadata database table)))
+  (cond-> (:fields (fetch-metadata/table-metadata database table))
+    (driver/supports? (:engine database) :nested-field-columns)
+    (clojure.set/union (fetch-metadata/nfc-metadata database table))))
```
Comment (dpsutton) [inline src/metabase/sync/sync_metadata/fields/fetch_metadata.clj:89]: don't use a fully qualified name like this. pull in set and use its alias.

File: src/metabase/sync/sync_metadata/fields/fetch_metadata.clj [hunk 3]
```diff
@@ -94,4 +84,6 @@
   "Fetch metadata about Fields belonging to a given `table` directly from an external database by calling its driver's
   implementation of `describe-table`."
   [database :- i/DatabaseInstance table :- i/TableInstance]
-  (:fields (fetch-metadata/table-metadata database table)))
+  (cond-> (:fields (fetch-metadata/table-metadata database table))
+    (driver/supports? (:engine database) :nested-field-columns)
```
Comment (dpsutton) [inline src/metabase/sync/sync_metadata/fields/fetch_metadata.clj:88]: use `driver/database-supports?` if possible here.

File: src/metabase/driver/postgres.clj [hunk 4]
```diff
@@ -342,14 +344,27 @@
     (pretty [_]
       (format "%s::%s" (pr-str expr) (name psql-type)))))
 
+(defn- json-query [identifier nfc-path]
+  (letfn [(handle-name [x] (if (number? x) (str x) (name x)))]
+    (apply hsql/call [:json_extract_path_text
+                      (hx/cast :json (keyword (first nfc-path)))
```
Comment (dpsutton) [inline src/metabase/driver/postgres.clj:350]: I apologize i haven't read the product docs, but i'm assuming based on this that the first element of the path is the underlying json field? If so, why are we casting to json?

**Review by dpsutton — APPROVED** [review]

File: src/metabase/driver/postgres.clj (hunk 4 above)
Comment (howonlee) [reply src/metabase/driver/postgres.clj:350]: We cast to json because it's not happy with jsonb type

**Pull request merged**